---
title: "RNA binding protein search"
output: html_document
date: "2022-08-05"
---

#FOURTH

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
library("data.table")
library("tidyverse")
library("pheatmap")
library("rex")
library("modeest")
library("dplyr")
library("plyr")
library("grid")
library("gridBase")
library("stringi")
library("RColorBrewer")
library("ggplot2")
library("purrr")
```

```{r}
load("C:/Users/xcwol/Downloads/.RData")
```


# Convert RBP names to WBGene:

# First I would like to create a file with all the RBPs of interest:
```{r}
RBP_test_genes_file <- read.csv(file = "HUGHESRBPgenenames.csv", header = T, row.names = 1)
RBP_test_genes <- rownames(RBP_test_genes_file)

mined_RBPs_file <- read.csv(file = "simplemine_results.csv", header = T) # Duplicates of nog-1: used the row that contained "public name" of nog-1
mined_RBPs <- mined_RBPs_file[,2:3]
RBP_list <- c(mined_RBPs[,1])
```

# Then, I would like to search the .csv files I made in type vs type 1 x 1 for these RBPs:
```{r}
setwd("D:/Zach Wolfe's DESeq analysis")

# This works. The select() function is the problem.
# select() function is for variables, not values (I think). We need to find a function for selecting values

overlap_data <- fread(file = "resAWBvsVC.csv", header = TRUE)[,c(1,3,7)]
match(RBP_list, overlap_data$V1, nomatch = NA)
print(length(which(match(RBP_list, overlap_data$V1) < 486 ))) # We know our RNABP list is 484 genes long. I need to add 2 to 484 (row 1 is header (+1), less than function adds +1) I am searching the list of ALL genes and printing the number of RNABPs expressed in the top 484 entries of each comparison. Not statistically accurate but we are getting somewhere.
```

# Then, I want to display these differential expression measures in a new table:
# Maybe use Microsoft Access or try to do this manually
```{r}
setwd("D:/Zach Wolfe's DESeq analysis")

RBP_results <- matrix(nrow = 41, ncol = 41, dimnames = list(singletypes,singletypes))
RBP_results[is.na(RBP_results)] <- 0

for(i in singletypes){
  for(j in singletypes){
    if(i != j){

overlap_data <- fread(file = paste("res", i, "vs", j, ".csv", sep=""))[,c(1,3,7)]
match(RBP_list, overlap_data$V1, nomatch = NA)

RBP_results[i,j] <- print(length(which(match(RBP_list, overlap_data$V1) < 312))) # 312 is top 1% significant of all genes (31173)

# RBP_results[i,j] <- write.csv(overlap_data, file=paste("RBPs", i, "vs", j, ".csv", sep="")) # Write csv of results

    }
    if(i == j) next
    }
}
```

```{r}
comparison <- print(paste(ADL, "vs", VD))
# gene <- print(overlap_data$V1[which(match(RBP_list, overlap_data$V1) < 312)])
# log2foldchange <- print(overlap_data$log2FoldChange[which(match(RBP_list, overlap_data$V1) < 312)])
# pvalue <- print(overlap_data$padj[which(match(RBP_list, overlap_data$V1) < 312)])
```

# Print names of RNABPs from chart above:
#STOP: Before you run this whole chunk, run line 116 JUST ONCE, then "#" it out. This is the easiest way to create the correctly formatted topDEgene object
```{r, include=FALSE}
setwd("D:/Zach Wolfe's DESeq analysis")

RBP_cutoff_results <- matrix(nrow = 41, ncol = 41, dimnames = list(singletypes,singletypes))
RBP_cutoff_results[is.na(RBP_results)] <- 0

for(i in singletypes){
   for(j in singletypes){
    if(i != j){

overlap_data <- fread(file = paste("res", i, "vs", j, ".csv", sep = ""))[,c(1,3,7)] %>% as.data.frame()

rownames(overlap_data) <- overlap_data$V1
RBP_list %in% rownames(overlap_data)
intersection <- rownames(overlap_data) %in% RBP_list
# print(overlap_data[intersection,])

comparison <- print(paste(i, "vs", j))
new_row <- print(paste(comparison, "#1"))
new_row2 <- print(paste(comparison, "#2"))
new_row3 <- print(paste(comparison, "#3"))

totaldata <- data.frame(overlap_data[intersection,])
totaldata <- na.omit(totaldata)
# topDEgene <- as.data.frame(totaldata[1,], row.names = comparison, col.names = c("V1", "log2FoldChange", "padj")) # run this AFTER the line above JUST ONCE, then "#" out this row. This is how we create the correctly formatted original topDEgene object
topDEgene[new_row,] <- rbind(totaldata[1,], new_row)
topDEgene[new_row2,] <- rbind(totaldata[2,], new_row2)
topDEgene[new_row3,] <- rbind(totaldata[3,], new_row3)
cutoffgenes <- as.data.frame(totaldata[which(abs(totaldata$log2FoldChange)>2 & totaldata$padj < .01),]) # Change cutoffs and abs() if desired
# topDEgene[new_row,] <- rbind(totaldata[which(abs(overlap_data$log2FoldChange)>2 & overlap_data$padj < .01),], new_row) # Only codes 1 row. WRONG
# RBP_cutoff_results[i,j] <- print(nrow(cutoffgenes))
# write.csv(cutoffgenes, file=paste("DiffExpRBPs", i, "vs", j, ".csv", sep=""))

# generate histograms using cutoffgenes:

#      if (nrow(cutoffgenes) > 0){
# cutoff_plot <- plot(cutoffgenes$log2FoldChange, y = cutoffgenes$padj, type = "p", xlim = c(-8,10), ylim = c(cutoffgenes[nrow(cutoffgenes),3],cutoffgenes[1,3]), log = "y", pch = 16, col = rainbow(15), main = paste("Differentially Experessed RBPs (WormBase ID) between", i, "and", j), xlab = "log2FoldChange value", ylab = "p-value (adj.)")
# text(cutoffgenes$log2FoldChange, cutoffgenes$padj, stri_sub(cutoffgenes$V1, 9, 14), cex = 1.5, pos = 4, col = rainbow(15))

# cutoff_barplot <- barplot(cutoffgenes$log2FoldChange, width = 0.8, col = rainbow(10), main = paste("Differentially Expressed RBPs between", i, "and", j), xlab = "log2FoldChange of Gene", ylab = "", horiz = TRUE)
# text(0, cutoff_barplot, cutoffgenes$V1, cex = 0.8, pos = 4)
#      }
#    if (nrow(cutoffgenes) == 0) next
     }
    if (i == j) next
   }
}

RBP_cutoff_results[is.na(RBP_cutoff_results)] <- 0
```

# Get rid of the first row in topDEgene (created in line 116):
```{r}
# topDEgene <- topDEgene[-1,] # run this AFTER the chunk above JUST ONCE, then "#" out this row
```

```{r}
write.csv(topDEgene[1641:6560,], file = "topDifferentiallyExpressedGenesbyComparison.csv")
write.csv(RBP_cutoff_results, file = "RBP_cutoff_results.csv")
```

# Backup in case something happens to the matrix above:
```{r}
#RBP_cutoff_results <- read.csv(file = "RBP_cutoff_results.csv", header = TRUE)
#rownames(RBP_cutoff_results) <- colnames(RBP_cutoff_results[2:42])
#RBP_cutoff_results <- RBP_cutoff_results[2:42]
```

```{r}
plot.new()
pheatmap(RBP_cutoff_results,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         cluster_rows = F,
         cluster_cols = F,
         fontsize = 7.5,
         col = rev(colors),
         main = "Number of Differentially Expressed RBPs by Cell Type (out of top 1% diff. expressed genes)")

plot.new()
pheatmap(RBP_cutoff_results,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         cluster_rows = F,
         cluster_cols = F,
         display_numbers = T,
         number_format = "%.0f",
         number_color = "black",
         fontsize = 7.5,
         fontsize_number = 6.75,
         col = colors,
         main = "Number of Differentially Expressed RBPs by Cell Type (out of top 1% diff. expressed genes)")
```

## (12-19-22) Redo this chunk by including 0-row data frames, then make a chunk below to save/reload the correct log2foldchange_per_gene:
```{r}
setwd("D:/Zach Wolfe's DESeq analysis")

log2foldchange_per_gene <-  matrix(nrow = 41, ncol = 484, dimnames = list(singletypes, mined_RBPs$WormBase.Gene.ID))
log2foldchange_per_gene[is.na(log2foldchange_per_gene)] <- 0

for(j in singletypes){
 for(k in singletypes){
  if(j != k){

DiffExpRBPs <- fread(file = paste("DiffExpRBPs", j, "vs", k, ".csv", sep = ""))[,2:3] %>% as.data.frame()
log2foldchange_per_gene[j,DiffExpRBPs$V1] <- print(DiffExpRBPs$log2FoldChange) + log2foldchange_per_gene[j, DiffExpRBPs$V1]
     }
      if (j == k) next
     }
   }

log2foldchange_per_gene_zeroes <- (colSums(log2foldchange_per_gene) == 0)
log2foldchange_per_gene <- log2foldchange_per_gene[, !log2foldchange_per_gene_zeroes]
```

# Which cell types have the most upregulation and downregulation?
```{r}
rownames(log2foldchange_per_gene)[rowSums(log2foldchange_per_gene) == max(rowSums(log2foldchange_per_gene))]
rownames(log2foldchange_per_gene)[rowSums(log2foldchange_per_gene) == min(rowSums(log2foldchange_per_gene))]

upreg_cell_type <- rownames(log2foldchange_per_gene)[rowSums(log2foldchange_per_gene) == max(rowSums(log2foldchange_per_gene))]
downreg_cell_type <- rownames(log2foldchange_per_gene)[rowSums(log2foldchange_per_gene) == min(rowSums(log2foldchange_per_gene))]

log2foldchange_per_gene[upreg_cell_type,]
log2foldchange_per_gene[downreg_cell_type,]
```

# Which genes have the greatest difference in regulation?
```{r}
log2foldchange_per_gene_colnames <- colnames(log2foldchange_per_gene)

for(k in 1:ncol(log2foldchange_per_gene)){
print(log2foldchange_per_gene_colnames[k])
print(max(log2foldchange_per_gene[,k]-min(log2foldchange_per_gene[,k])))
}
```

```{r}
colors <- colorRampPalette(brewer.pal(6, "Greens"))(250)
pheatmap(log2foldchange_per_gene,
         border_color = "black",
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         cluster_rows = F,
         cluster_cols = F,
         fontsize = 6,
         angle_col = 45,
         col = colors,
         main = "Total log2FoldChange Accross All Comparisons")

pheatmap(log2foldchange_per_gene[,1:50],
         border_color = "black",
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         cluster_rows = F,
         cluster_cols = F,
         fontsize = 7,
         angle_col = 90,
         col = colors,
         main = "Total log2FoldChange Accross All Comparisons (first quintile)")

pheatmap(log2foldchange_per_gene[,51:100],
         border_color = "black",
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         cluster_rows = F,
         cluster_cols = F,
         fontsize = 7,
         angle_col = 90,
         col = colors,
         main = "Total log2FoldChange Accross All Comparisons (second quintile)")

pheatmap(log2foldchange_per_gene[,101:150],
         border_color = "black",
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         cluster_rows = F,
         cluster_cols = F,
         fontsize = 7,
         angle_col = 90,
         col = colors,
         main = "Total log2FoldChange Accross All Comparisons (third quintile)")

pheatmap(log2foldchange_per_gene[,151:200],
         border_color = "black",
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         cluster_rows = F,
         cluster_cols = F,
         fontsize = 7,
         angle_col = 90,
         col = colors,
         main = "Total log2FoldChange Accross All Comparisons (fourth quintile)")

pheatmap(log2foldchange_per_gene[,201:257],
         border_color = "black",
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         cluster_rows = F,
         cluster_cols = F,
         fontsize = 7,
         angle_col = 90,
         col = colors,
         main = "Total log2FoldChange Accross All Comparisons (fifth quintile)")
```

# write.csv for sum of log2foldchange per gene (in which abs(log2FoldChange) > 2 & padj < .01) in each cell type) - colSums(0) columns removed
```{r}
write.csv(log2foldchange_per_gene, file = "sum of log2foldchange per gene.csv")
```

# Backup read.csv:
```{r}
log2foldchange_per_gene <- read.csv("sum of log2foldchange per gene.csv", row.names = 1)
```


# Combine every excel file for each cell type:

# Note: If there are 0 Diff. Exp. RBPs in a comparison, dplyr will assume that the rows in the .csv file are logical (we want them to be character/numeric). This will prevent us from merging rows in different .csv files together in the upcoming chunk. You may have to examine each Diff. Exp. RBP file and manually move them outside of your working directory if there are 0 rows:

## cell types with 0 Diff. Exp. RBPs in my example can be found in D:\Zach Wolfe's DESeq analysis\Zero_Diff_Exp_RBPs_comparisons 
```{r}
setwd("D:/Zach Wolfe's DESeq analysis")

ADLDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsADLvs*")
AFDDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsAFDvs*")
AINDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsAINvs*")
AIYDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsAIYvs*")
AINDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsAINvs*")
ASELDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsASELvs*")
ASERDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsASERvs*")
ASGDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsASGvs*")
ASIDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsASIvs*")
ASKDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsASKvs*")
AVADiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsAVAvs*")
AVEDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsAVEvs*")
AVGDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsAVGvs*")
AVHDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsAVHvs*")
AVKDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsAVKvs*")
AVMDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsAVMvs*")
AWADiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsAWAvs*")
AWBDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsAWBvs*")
AWCDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsAWCvs*")
BAGDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsBAGvs*")
CANDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsCANvs*")
DADiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsDAvs*")
DDDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsDDvs*")
I5DiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsI5vs*")
IL1DiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsIL1vs*")
IL2DiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsIL2vs*")
NSMDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsNSMvs*")
OLLDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsOLLvs*")
OLQDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsOLQvs*")
PHADiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsPHAvs*")
PVCDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsPVCvs*")
PVDDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsPVDvs*")
PVMDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsPVMvs*")
RIADiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsRIAvs*")
RICDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsRICvs*")
RIMDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsRIMvs*")
RISDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsRISvs*")
RMDDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsRMDvs*")
SMDDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsSMDvs*")
VBDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsVBvs*")
VCDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsVCvs*")
VDDiffExpRBPfiles <- list.files(pattern = "DiffExpRBPsVDvs*")
```

# Make a list of all DiffExpRBPfiles for each cell type:
```{r}
DiffExpRBPfileList <- rep(paste0(singletypes, "DiffExpRBPfiles"), len = 41)
```


# Log2FoldChange values for unique RBP names in individual "cell type vs" files:
```{r}
setwd("D:/Zach Wolfe's DESeq analysis")

LargeRBPdf <- AFDDiffExpRBPfiles %>% map_dfr(read.csv) # change cell type here
LargeRBPdf <- LargeRBPdf[,-c(1)]
uniqueRBPs <- unique(LargeRBPdf$V1)

logSums <- matrix(nrow = length(uniqueRBPs), ncol = 1, dimnames = list(uniqueRBPs), log2foldchange)
logSums[is.na(logSums)] <- 0

logMeans <- matrix(nrow = length(uniqueRBPs), ncol = 1, dimnames = list(uniqueRBPs), log2foldchange)
logMeans[is.na(logMeans)] <- 0

for(i in uniqueRBPs){
logSums[i,1] <- sum(LargeRBPdf[LargeRBPdf$V1 == i, 2]) # perhaps mean would be more appropriate than sum for this type of comparison?
logMeans[i,1] <- mean(LargeRBPdf[LargeRBPdf$V1 == i, 2])
}
```

# Save/change cell type here:
```{r}
write.csv(logSums, file = "logSumsAFD.csv")
write.csv(logMeans, file = "logMeansAFD.csv") # change cell type here
#write.csv(LargeRBPdf, file = "Most interesting RBPs in cell type AFD.csv")
```

# Make a heatmap for the logSums value in each gene for a given cell type:
```{r}
logSums_table <- matrix(nrow = 484, ncol = 41, dimnames = list(RBP_list,singletypes))
logSums_table[is.na(logSums_table)] <- 0

for (i in singletypes){
logSums_for_graph <- fread(file = paste0("logSums", i, ".csv", sep = "")) %>% as.data.frame()
rownames(logSums_for_graph) <- logSums_for_graph[,1]
  for (j in RBP_list){
logSums_table[j,i] <- print(logSums_for_graph[j,2])
logSums_table[is.na(logSums_table)] <- 0
  }
}

colors <- colorRampPalette(brewer.pal(6, "PuRd"))(250)
pheatmap(logSums_table,
         border_color = "black",
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         cluster_rows = F,
         cluster_cols = F,
         display_numbers = F,
         number_format = "%.2f",
         number_color = "black",
         fontsize = 6,
         fontsize_number = 6,
         angle_col = 45,
         col = colors,
         main = "Sum of All log2FoldChange Values for Unique RBPs in One Cell Type vs All Other Cell Types")

logSums_table <- logSums_table[rowSums(logSums_table) != 0,] # if you want to keep rows with 0 data, "#" this row out
maxDEcelltypes <- count(colnames(logSums_table)[apply(logSums_table, 1, which.max)])
maxDEcelltypes <- maxDEcelltypes[order(maxDEcelltypes$freq, decreasing = TRUE),]
minDEcelltypes <- count(colnames(logSums_table)[apply(logSums_table, 1, which.min)])
minDEcelltypes <- minDEcelltypes[order(minDEcelltypes$freq, decreasing = TRUE),]
maxDERBPs <- count(colnames(t(logSums_table))[apply(t(logSums_table), 1, which.max)])
maxDERBPs <- maxDERBPs[order(maxDERBPs$freq, decreasing = TRUE),]
minDERBPs <- count(colnames(t(logSums_table))[apply(t(logSums_table), 1, which.min)])
minDERBPs <- minDERBPs[order(minDERBPs$freq, decreasing = TRUE),]
```

```{r}
write.csv(logSums_table, file = "logSums_table.csv")
write.csv(maxDEcelltypes, file = "maxDEcelltypes.csv")
write.csv(minDEcelltypes, file = "minDEcelltypes.csv")
write.csv(maxDERBPs, file = "maxDERBPs.csv")
write.csv(minDERBPs, file = "minDERBPs.csv")
```

# Backup:
```{r}
logSums_table <- read.csv("logSums_table.csv", row.names = 1)
```


# Which RBPs have the highest value overall in a single cell?
# Report: which RBP, what is the value, and which cell?
```{r}
log2foldchange_summary_df <- data.frame()

#for (i in singletypes){
#log2foldchange_summary_df[i,1] <- print(rownames(logSums_table)[which.max(logSums_table[,i])])
#log2foldchange_summary_df[i,2] <- print(max(logSums_table[,i]))
#}

for (i in singletypes){
cell_type <- print(paste(i))
new_row <- print(paste(cell_type, "#1"))
new_row2 <- print(paste(cell_type, "#2"))
new_row3 <- print(paste(cell_type, "#3"))
new_row4 <- print(paste(cell_type, "#4"))
new_row5 <- print(paste(cell_type, "#5"))
new_row6 <- print(paste(cell_type, "#6"))
new_row7 <- print(paste(cell_type, "#7"))
new_row8 <- print(paste(cell_type, "#8"))
new_row9 <- print(paste(cell_type, "#9"))
new_row10 <- print(paste(cell_type, "#10"))

logSums_table_new <- logSums_table[order(logSums_table[,i], decreasing = TRUE),]
log2foldchange_summary_df[new_row, 1] <- print(rownames(logSums_table_new)[1])
log2foldchange_summary_df[new_row2, 1] <- print(rownames(logSums_table_new)[2])
log2foldchange_summary_df[new_row3, 1] <- print(rownames(logSums_table_new)[3])
log2foldchange_summary_df[new_row4, 1] <- print(rownames(logSums_table_new)[4])
log2foldchange_summary_df[new_row5, 1] <- print(rownames(logSums_table_new)[5])
log2foldchange_summary_df[new_row6, 1] <- print(rownames(logSums_table_new)[6])
log2foldchange_summary_df[new_row7, 1] <- print(rownames(logSums_table_new)[7])
log2foldchange_summary_df[new_row8, 1] <- print(rownames(logSums_table_new)[8])
log2foldchange_summary_df[new_row9, 1] <- print(rownames(logSums_table_new)[9])
log2foldchange_summary_df[new_row10, 1] <- print(rownames(logSums_table_new)[10])

log2foldchange_summary_df[new_row, 2] <- print(logSums_table_new[1,i])
log2foldchange_summary_df[new_row2, 2] <- print(logSums_table_new[2,i])
log2foldchange_summary_df[new_row3, 2] <- print(logSums_table_new[3,i])
log2foldchange_summary_df[new_row4, 2] <- print(logSums_table_new[4,i])
log2foldchange_summary_df[new_row5, 2] <- print(logSums_table_new[5,i])
log2foldchange_summary_df[new_row6, 2] <- print(logSums_table_new[6,i])
log2foldchange_summary_df[new_row7, 2] <- print(logSums_table_new[7,i])
log2foldchange_summary_df[new_row8, 2] <- print(logSums_table_new[8,i])
log2foldchange_summary_df[new_row9, 2] <- print(logSums_table_new[9,i])
log2foldchange_summary_df[new_row10, 2] <- print(logSums_table_new[10,i])
}

colnames(log2foldchange_summary_df) = c("RBP", "sum of log2foldchange")
```

```{r}
write.csv(log2foldchange_summary_df, file = "which_RBP_max_per_cell_type.csv")
```

# Now let's find which RBPs are the most differentially expressed accross ALL cell types, not just in individual cells or comparisons:
```{r}
log2foldchange_per_gene_colsums <- colSums(log2foldchange_per_gene)
log2foldchange_per_gene_colsums <- log2foldchange_per_gene_colsums[order(log2foldchange_per_gene_colsums, decreasing = TRUE)]
log2foldchange_per_gene_colsums <- log2foldchange_per_gene_colsums[order(log2foldchange_per_gene_colsums, decreasing = FALSE)]
```

```{r}
write.csv(log2foldchange_per_gene_colsums, file = "log2foldchange of all 484 RBPs.csv")
```

# Looks like WBGene00010046 and WBGene00016260 are the "most intersting" for now. Let's examine them each more closely:
```{r}
colors <- colorRampPalette(brewer.pal(20, "PuRd"))(250)
interesting_RBP_table <- logSums_table[c("WBGene00010046", "WBGene00016260"),]

pheatmap(interesting_RBP_table,
         border_color = "black",
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         cluster_rows = F,
         cluster_cols = F,
         show_rownames = T,
         show_colnames = T,
         display_numbers = F,
         fontsize = 7.5,
         angle_col = 45,
         col = colors,
         main = "Sum of All log2FoldChange Values for Interesting RBPs in One Cell Type vs All Other Cell Types")

interesting_RBP_table <- logSums_table[c("WBGene00010046", "WBGene00016260", "WBGene00022771", "WBGene00019347", "WBGene00006321", "WBGene00022250"),]

pheatmap(interesting_RBP_table,
         border_color = "black",
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         cluster_rows = F,
         cluster_cols = F,
         show_rownames = T,
         show_colnames = T,
         display_numbers = F,
         fontsize = 7.5,
         angle_col = 45,
         col = colors,
         main = "Sum of All log2FoldChange Values for Interesting RBPs in One Cell Type vs All Other Cell Types")

pheatmap(logSums_table[,c("AFD", "ASEL", "ASER")],
         border_color = "black",
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         cluster_rows = F,
         cluster_cols = F,
         show_rownames = F,
         show_colnames = T,
         display_numbers = F,
         fontsize = 7,
         angle_col = 45,
         col = colors,
         main = "Sum of All log2FoldChange Values for AFD, ASEL, and ASER vs All Other Cell Types")
```


# This is an attempt to loop several chunks above, but does not work as of 12-6-22:
```{r}
setwd("D:/Zach Wolfe's DESeq analysis")

for(j in DiffExpRBPfileList){
  for(k in singletypes){

LargeRBPdf <- cat(j, "\n") %>% map_dfr(read.csv) # can't open the connection??
#LargeRBPdf <- LargeRBPdf[,-c(1)]
uniqueRBPs <- unique(LargeRBPdf[[V1]]) #can't find column $V1, and can't subset columns using [,1] #because LargeRBPdf is empty!

logSums <- matrix(nrow = length(uniqueRBPs), ncol = 1, dimnames = list(uniqueRBPs), log2foldchange)
logSums[is.na(logSums)] <- 0

logMeans <- matrix(nrow = length(uniqueRBPs), ncol = 1, dimnames = list(uniqueRBPs), log2foldchange)
logMeans[is.na(logMeans)] <- 0

for(i in uniqueRBPs){
logSums[i,1] <- sum(LargeRBPdf[LargeRBPdf$V1 == i, 2]) # perhaps mean would be more appropriate than sum for this type of comparison?
logMeans[i,1] <- mean(LargeRBPdf[LargeRBPdf$V1 == i, 2])
}
write.csv(logMeans, file = paste0("RBP_logMeans_for_", k, ".csv"))}
}
```

# Top Differentially Expressed ("topDE") genes for each cell type:
```{r}
uniqueDERBPs <- c(unique(topDEgene$V1))
# only2topDEgenes <- c("WBGene00000105", "WBGene00001094", "WBGene00001340", "WBGene00003026", "WBGene00003218", "WBGene00004418", "WBGene00004472", "WBGene00004704", "WBGene00004795", "WBGene00004945", "WBGene00006698", "WBGene00008224", "WBGene00008561", "WBGene00013307", "WBGene00016113", "WBGene00018612", "WBGene00019971", "WBGene00021901", "WBGene00021929", "WBGene00022029")

cnt <- count(topDEgene$V1) # CRITICAL: counts of unique D.E. RBPs
cnt_ordered <- cnt[order(cnt$freq),] # ordered counts of unique D.E. RBPs
twotopDEgenes <- cnt[which(cnt$freq == 2),] # CRITICAL: all unique D.E. RBPs that only show as max in 1 comparison
cnt[which.max(cnt$freq),] # max top D.E. gene across all comparisons

ADLuniqueRBPs <- topDEgene[1:40, 1:3]
ADLs <- sapply(ADLuniqueRBPs, function(c){
  sapply(ADLuniqueRBPs$V1, function(y){
    sum(grepl(y, c))
  })
})
unique(rownames(ADLs[ADLs[,1] == max(ADLs[,1]),])) # gives the gene that most frequently occurs in comparisons

AFDuniqueRBPs <- topDEgene[41:80, 1:3]
AFDs <- sapply(AFDuniqueRBPs, function(c){
  sapply(AFDuniqueRBPs$V1, function(y){
    sum(grepl(y, c))
  })
})
unique(rownames(AFDs[AFDs[,1] == max(AFDs[,1]),])) # gives the gene that most frequently occurs in comparisons

AINuniqueRBPs <- topDEgene[81:120, 1:3]
AINs <- sapply(AINuniqueRBPs, function(c){
  sapply(AINuniqueRBPs$V1, function(y){
    sum(grepl(y, c))
  })
})
unique(rownames(AINs[AINs[,1] == max(AINs[,1]),])) # gives the gene that most frequently occurs in comparisons

AIYuniqueRBPs <- topDEgene[121:160, 1:3]
AIYs <- sapply(AIYuniqueRBPs, function(c){
  sapply(AIYuniqueRBPs$V1, function(y){
    sum(grepl(y, c))
  })
})
unique(rownames(AIYs[AIYs[,1] == max(AIYs[,1]),])) # gives the gene that most frequently occurs in comparisons

ASELuniqueRBPs <- topDEgene[161:200, 1:3]
ASELs <- sapply(ASELuniqueRBPs, function(c){
  sapply(ASELuniqueRBPs$V1, function(y){
    sum(grepl(y, c))
  })
})
unique(rownames(ASELs[ASELs[,1] == max(ASELs[,1]),])) # gives the gene that most frequently occurs in comparisons

ASERuniqueRBPs <- topDEgene[201:240,1:3]
ASERs <- sapply(ASERuniqueRBPs, function(c){
  sapply(ASERuniqueRBPs$V1, function(y){
    sum(grepl(y, c))
  })
})
unique(rownames(ASERs[ASERs[,1] == max(ASERs[,1]),])) # gives the gene that most frequently occurs in comparisons

OLLuniqueRBPs <- topDEgene[1041:1080,1:3]
OLLs <- sapply(OLLuniqueRBPs, function(c){
  sapply(OLLuniqueRBPs$V1, function(y){
    sum(grepl(y, c))
  })
})
unique(rownames(OLLs[OLLs[,1] == max(OLLs[,1]),])) # gives the gene that most frequently occurs in comparisons
```

# Loop the previous chunk:
```{r}
typesplusuniqueRBPs <- c("ADLuniqueRBPs","AFDuniqueRBPs","AINuniqueRBPs","AIYuniqueRBPs","ASELuniqueRBPs","ASERuniqueRBPs","ASGuniqueRBPs","ASIuniqueRBPs","ASKuniqueRBPs","AVAuniqueRBPs","AVEuniqueRBPs","AVGuniqueRBPs","AVHuniqueRBPs","AVKuniqueRBPs","AVMuniqueRBPs","AWAuniqueRBPs","AWBuniqueRBPs","AWCuniqueRBPs","BAGuniqueRBPs","CANuniqueRBPs","DAuniqueRBPs","DDuniqueRBPs","I5uniqueRBPs","IL1uniqueRBPs","IL2uniqueRBPs","NSMuniqueRBPs","OLLuniqueRBPs","OLQuniqueRBPs","PHAuniqueRBPs","PVCuniqueRBPs","PVDuniqueRBPs","PVMuniqueRBPs","RIAuniqueRBPs","RICuniqueRBPs","RIMuniqueRBPs","RISuniqueRBPs","RMDuniqueRBPs","SMDuniqueRBPs","VBuniqueRBPs","VCuniqueRBPs","VDuniqueRBPs")

for (i in typesplusuniqueRBPs){
  for (j in singletypes){

i <- topDEgene[grep(paste("^", j, sep = ""), rownames(topDEgene)),]
japply <- sapply(i, function(c){
  sapply(i[,1], function(y){
    sum(grepl(y, c))
  })
})
print(j)
print(unique(rownames(japply[japply[,1] == max(japply[,1]),]))) # gives gene that most frequently occurs as #1 most diff. expressed in comparisons
print(nrow(japply[japply[,1] == max(japply[,1]),])) # gives the number of times that the gene above is #1 most diff. expressed in comparisons
  }
}
```

# Plots:
```{r}
plot.new()
pheatmap(RBP_cutoff_results,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         cluster_rows = F,
         cluster_cols = F,
         fontsize = 7.5,
         col = colorRampPalette(brewer.pal(9, "Blues"))(250), # Can we color this differenly by motor vs sensory vs inter?
         main = "Number of Differentially Expressed RBPs by Cell Type (log2FoldChange > 2 and adj. p-value < 0.01)")
```

```{r}
plot.new()
cnt_barplot <- barplot((as.numeric(cnt_ordered[111:130,2])/2), beside = TRUE, names.arg = "", angle = 85, col = rainbow(35), main = "Top 20 Differentially Expressed RBPs (WormBase ID) among All Cell Types", xlab = "", ylab = "Count", plot = TRUE) # need to divide by 2: XXX is compared to YYY and also YYY is compared to XXX - these will yield the same result
vps <- baseViewports()
pushViewport(vps$inner, vps$figure, vps$plot)
grid.text(stri_sub(cnt_ordered[111:130,1], 9, 14),
           x = unit(cnt_barplot, "native"), y=unit(-1.5, "lines"),
    just="bottom", rot=80)
popViewport(3)

barplot(as.numeric((cnt_ordered[31:130,2])/2) ~ stri_sub(cnt_ordered[31:130,1], 9, 14), # need to divide by 2: XXX is compared to YYY and also YYY is compared to XXX - these will yield the same result
        main = "How Often is an RBP in the Top 3 D.E. for a Cell Type?",
        xlab = "", ylab = "Count",
        col = rainbow(40),
        las=2,
        cex.names=0.65,
        plot = TRUE)
```

```{r}
ggplot(cnt_ordered, aes(x = x, y = as.numeric(freq/2), show.legend = TRUE)) +
  geom_bar(stat = "identity", color = rainbow(130), fill = rainbow(130)) +
  labs(title= "How Often is an RBP in the Top 3 D.E. for a Cell Type?", 
         x = "", y = "Count")
  theme_minimal()
```

```{r}
AINplot <- barplot(table(AINuniqueRBPs$V1),
        beside = TRUE,
        main = "How Often is an RBP the Most Differentially Expressed in a Comparison with AIN?",
        ylab = "Count",
        col = rainbow(25),
        las=2,
        cex.names = 0.65,
        plot = TRUE)
text(x = AINplot, y = as.numeric(table(AINuniqueRBPs$V1)), label = as.numeric(table(AINuniqueRBPs$V1)), cex = 1, pos = 1, col = "black")

OLLplot <- barplot(table(OLLuniqueRBPs$V1),
        beside = TRUE,
        main = "How Often is an RBP the Most Differentially Expressed in a Comparison with OLL?",
        ylab = "Count",
        col = rainbow(25),
        las=2,
        cex.names = 0.5,
        plot = TRUE)
text(x = OLLplot, y = as.numeric(table(OLLuniqueRBPs$V1)), label = as.numeric(table(OLLuniqueRBPs$V1)), cex = 1, pos = 1, col = "black")
```

# Attempt at ggplot bubble plot:
```{r}
ggplot(AINuniqueRBPs, aes(x=rev(padj), y=V1, size=log2FoldChange)) +
  geom_point(aes(col = "blue")) +
  theme(axis.text.x = element_blank())
```
